src/XrdSciTokens/XrdSciTokensAccess.cc

```diff
@@ -862,20 +862,43 @@ class XrdAccSciTokens : public XrdAccAuthorize, public XrdSciTokensHelper,
         int idx = 0;
         std::set<std::string> paths_write_seen;
         std::set<std::string> paths_create_or_modify_seen;
+        std::vector<std::string> acl_paths;
+        acl_paths.reserve(config.m_restricted_paths.size() + 1);
         while (acls[idx].resource && acls[idx++].authz) {
+            acl_paths.clear();
             const auto &acl_path = acls[idx-1].resource;
             const auto &acl_authz = acls[idx-1].authz;
-            if (!config.m_restricted_paths.empty()) {
-                bool found_path = false;
+            if (config.m_restricted_paths.empty()) {
+                acl_paths.push_back(acl_path);
+            } else {
+                auto acl_path_size = strlen(acl_path);
                 for (const auto &restricted_path : config.m_restricted_paths) {
+                    // See if the acl_path is more specific than the restricted path; if so, accept it
+                    // and move on to applying paths.
                     if (!strncmp(acl_path, restricted_path.c_str(), restricted_path.size())) {
-                        found_path = true;
+                        // Only do prefix checking on full path components.  If acl_path=/foobar and
+                        // restricted_path=/foo, then we shouldn't authorize access to /foobar.
+                        if (acl_path_size > restricted_path.size() && acl_path[restricted_path.size()] != '/') {
+                            continue;
+                        }
+                        acl_paths.push_back(acl_path);
                         break;
                     }
+                    // See if the restricted_path is more specific than the acl_path; if so, accept the
+                    // restricted path as the ACL.  Keep looping to see if other restricted paths add
+                    // more possible authorizations.
+                    if (!strncmp(acl_path, restricted_path.c_str(), acl_path_size)) {
+                        // Only do prefix checking on full path components.  If acl_path=/foo and
+                        // restricted_path=/foobar, then we shouldn't authorize access to /foobar.
+                        if (restricted_path.size() > acl_path_size && restricted_path[acl_path_size] != '/') {
```

Comment:
pplSame here, but for `acl_path` = `/store` and `restricted_path` = `/store/`. I think there should be a function factored out that checks if a directory is a subdirectory of another, and use it in both places. You may also want to think about symlinks which may let a user bypass the restrictions (i.e. a symlink in the user directory under the restricted path which points to `/` and may let the user read more than they should).

Reply:
same here restricted_path[acl_path_size-1] != '/'

Reply:
Actually, I think the check is correct, since it checks that for e.g. `acl_path = /store` and `restricted_path = /store/cms` it works and for `restricted_path = /storecms` it doesn't. That is, the `/` character is at position one bigger than the size of `acl_path`, which is correct.

Reply:
the patch I put in would still reject `/store` if restricted_path is `/storecms`. it would pass the stringcompare and then correctly detect the last character as `e`, rejecting it at the  `!= '/'` condition.
The case I had to put in the -1 for was when the acl_path is `/` and restricted path is `/store`, which would still fail even if you change the strcmp condition

Reply:
restricted_path is already 0 indexed, so the last character's position would be at `acl_path_size-1`
in fact, with the current condition, an `acl_path = /store/` would fail a check with `restricted_path= /store/cms` for the same reason, while with the -1 it would pass and add the restricted_path

Reply:
> You may also want to think about symlinks which may let a user bypass the restrictions (i.e. a symlink in the user directory under the restricted path which points to / and may let the user read more than they should).

Symlinks are a whole other set of (valid) concerns. For example, think about how the `all.export` setting works if you include user-controlled symlinks.  I do think it's a conversation worth having but at a higher level -- I don't want to derail this PR with a larger orthogonal issue.